src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+
+function maybeIntern(value:(RawValue|Register)):Register|ID {
+  if(value === undefined || value === null) throw new Error("Trying to intern an undefined");
+  if(isRegister(value)) return value;
+  return GlobalInterner.intern(value);
+}
+
+//--------------------------------------------------------------------
+// DSLVariable
+//--------------------------------------------------------------------
+
+type DSLValue = RawValue|Register;
+class DSLVariable {
+  static CURRENT_ID = 0;
+  id: number;
+  constructor(public name:string, public value:DSLValue = new Register(-1)) {
+    this.id = DSLVariable.CURRENT_ID++;
+  }
+}
+
+//--------------------------------------------------------------------
+// DSLFunction
+//--------------------------------------------------------------------
+
+class DSLFunction {
+  returnValue:DSLVariable;
+
+  constructor(public block:DSLBlock, public path:string[], public args:any[]) {
+    this.returnValue = new DSLVariable("record");
+    block.registerVariable(this.returnValue);
+  }
+
+  compile() {
+    let constraints:FunctionConstraint[] = [];
+    let result = maybeIntern(this.block.toValue(this.returnValue));
+    let values = this.args.map((v) => maybeIntern(this.block.toValue(v)))
+    constraints.push(FunctionConstraint.create(this.path.join("/"), {result}, values) as FunctionConstraint);
+    return constraints;
+  }
+}
+
+//--------------------------------------------------------------------
+// DSLRecord
+//--------------------------------------------------------------------
+
+class DSLRecord {
+  record: DSLVariable;
+  output: boolean;
+  fields: any;
+  constructor(public block:DSLBlock, public tag:string[], initial:any) {
+    let fields:any = {tag};
+    for(let field in initial) {
+      let value = initial[field];
+      if(field.constructor !== Array) {
+        value = [value];
+      }
+      fields[field] = value;
+    }
+    this.fields = fields;
+    this.record = new DSLVariable("record");
+    block.registerVariable(this.record);
+  }
+
+  proxy() {
+    return new Proxy(this, {
+      get: (obj:any, prop:string) => {
+        if(obj[prop]) return obj[prop];
+        let found = obj.fields[prop];
+        if(prop === Symbol.toPrimitive) return () => {
+          return "uh oh";
+        }
+        if(!found) {
+          found = new DSLVariable(prop);
+          obj.fields[prop] = [found];
+          this.block.registerVariable(found);
+        } else {
+          found = found[0];
+        }
+        return found;
+      },
+      set: (obj:any, prop:string, value:any) => {
+        if(prop === "output") {
+          obj[prop] = value;
+          return true;
+        }
+        if(!obj.fields[prop]) {
+          if(value.constructor !== Array) {
+            value = [value];
+          }
+          obj.fields[prop] = value;
+          return true;
+        }
+        this.block.equivalence(obj.fields[prop][0], value);
```

Comment:
What does this array contain? Is it a tuple tracking some metadata rather than a list?